dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java

```diff
@@ -664,6 +665,10 @@ else if ( valueType.isBoolean() )
         {
             return false;
         }
+        else if ( valueType.isDate() )
```

Comment:
@jimgrace I needed to put a default value for dates because in the ANTLR parser library it changed the way we cast dates and now an empty string is throwing an exception.
What do you think is a good default value and does it make sense to fix it this way?

Reply:
Ha. Thanks for asking! How about we just return null and don't replace it with anything? Would that work?

When we insert a dummy value this is used as a real value returned to the user if an expression skip strategy is NEVER_SKIP, or if it is SKIP_IF_ALL_VALUES_MISSING and there is some other value present. With numbers, we can put in 0 as a reasonable default for a missing value. With booleans, we can put in false, and any test won't find the value to be true. But with dates no matter what value we substitute it's pretty safe to assume that it won't do what the user wants.

It would be better to say that if a date value is missing, the NEVER_SKIP or SKIP_IF_ALL_VALUES_MISSING strategies are overridden and no value is returned. If the user doesn't want this, they can always explicitly supply a default date value with a construct like firstNonNull( date, default ). That's better than us supplying a default date that is pretty certain to be what the user doesn't want.

Reply:
Hey @jimgrace ,
can you have another look?
I am still returning new Date() as default value as an actual date is needed to correctly compute the description.
But then, when evaluating, I am bypassing the null value strategy and if a null date is present in the evaluation, the whole expression is invalid

Reply:
Good point. Your fix looks perfect!

Note that you only set `nullDateFound = true` within the code block for `if ( replaceNulls )`. This block is not executed if you are inside functions like `isNull()`, `isNotNull()` or `firstNonNull()`. This means the user can still use these functions to test for a null date value in a controlled way (and supply a default if that is meaningful), and the expression can still return a value.

You might consider renaming `nullDateFound` to something like `unprotectedNullDateFound` to reflect this. Or not. :)